core/domain/user_jobs_one_off.py

```diff
@@ -150,6 +152,123 @@ def reduce(userbio_length, stringified_usernames):
             yield (userbio_length, stringified_usernames)
 
 
+class PopulateStoriesAndTopicsInIncompleteActivitiesOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to populate the story_ids and partially_learnt_topic_ids
+    in Incomplete Activities Model."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        """Return a list of datastore class references to map over."""
+        return [user_models.UserSettingsModel]
+
+    @staticmethod
+    def map(item):
+        """Implements the map function for this job."""
+        user_id = item.id
+        activity_ids_in_learner_dashboard = (
+            learner_progress_services.get_learner_dashboard_activities(
+                user_id))
+        incomplete_exploration_ids = (
+            activity_ids_in_learner_dashboard.incomplete_exploration_ids)
+        completed_exploration_ids = (
+            activity_ids_in_learner_dashboard.completed_exploration_ids)
+
+        story_ids_linked_to_incomplete_explorations = (
+            list(set(exp_services.get_story_ids_linked_to_explorations(
+                incomplete_exploration_ids))))
+        stories_linked_to_incomplete_explorations = (
+            story_fetchers.get_stories_by_ids(
+                story_ids_linked_to_incomplete_explorations))
+
+        story_ids_linked_to_completed_explorations = (
+            list(set(exp_services.get_story_ids_linked_to_explorations(
+                completed_exploration_ids))))
+        stories_linked_to_completed_explorations = (
+            story_fetchers.get_stories_by_ids(
+                story_ids_linked_to_completed_explorations))
+
+        # Mark story and topic as incomplete for incomplete explorations.
+        for story in stories_linked_to_incomplete_explorations:
+            if story:
+                learner_progress_services.mark_story_as_incomplete(
+                    user_id, story.id)
+                learner_progress_services.mark_topic_as_partially_learnt(
```

Comment:
A topic is fully learnt when at least one story is fully learnt.

I know that currently we only have one story per topic. But perhaps worth generalizing this to be fully correct, so that it can be used as a regeneration job anytime if data gets out of sync?

Reply:
I am not sure I understand this. The topic is marked as learnt when one story inside it is completed as we have only one story in a topic currently in the server. Here the topic is marked as incomplete because the exploration in the story is incomplete.

Reply:
A topic may, in general, have more than one story. What is your definition of "learnt" and "partially learnt" in the general case?

In general, do not write code based on what's currently on the server; what's on the server can change.

Reply:
Learnt means that all the stories in the topic are completed and partially learnt refers to at least one story has been started in it.

Reply:
Somehow my comment got lost, but: a topic is considered "learnt" if **at least one canonical story** in the topic has been completed. Could you please fix the logic here and elsewhere?

Reply:
Done.